bpf: fix asymmetric routing and cilium_host connectivity in v6 tunnel

borkmann opened the pull request:
(See commit description. This gets nodeport fully working in v6 tunneling mode.)

Fixes: #8654

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9062)
<!-- Reviewable:end -->

borkmann committed:
bpf: fix asymmetric routing and cilium_host connectivity in v6 tunnel mode

The current v6 + tunnel behavior is broken in multiple ways:

1) When one host tries to connect to a remote pod, then we find the
   destination address of the pod via ipcache lookup and given the
   info->tunnel_endpoint is set, we route the request via tunnel.

   BUT: with a cilium_host source IP which is a public address. This
   means that the reply of the pod is routed asymmetrically via public
   interface but not via tunnel!

   Some more context and background: for v6 (direct routing or tunnel
   mode), the cilium_host device has a globally scoped v6 address which
   matches the one of the public interface. This is due to the fact that
   we cannot add a private address with link scope like in v4 case where
   kernel can ensure that no traffic is leaked to the public:

   # ip a
   [...]
   2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    inet6 2a02:120b:2c12:c120:ec4:7aff:fa13:f90f/64 scope global dynamic mngtmpaddr noprefixroute
       valid_lft 262sec preferred_lft 22sec
   [...]
   12: cilium_host@cilium_net: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet6 2a02:120b:2c12:c120:ec4:7aff:fa13:f90f/128 scope global
       valid_lft forever preferred_lft forever
   [...]

   # ip -6 r
   [...]
   2001:db1:3::d899 dev cilium_host metric 1024 pref medium
   2001:db1:3::/112 via 2001:db1:3::d899 dev cilium_host src 2a02:120b:2c12:c120:ec4:7aff:fa13:f90f metric 1024 mtu 1450 pref medium
   2001:db8:5::/112 via 2001:db1:3::d899 dev cilium_host src 2a02:120b:2c12:c120:ec4:7aff:fa13:f90f metric 1024 mtu 1450 pref medium
   [...]

   This means that cilium_host effectively has two v6 addresses: HOST_IP
   which above is 2a02:120b:2c12:c120:ec4:7aff:fa13:f90f and ROUTER_IP
   from our v6 allocation pool which is 2001:db1:3::d899.

2) When one host tries to connect to a cidr address that is routable
   via tunnel, but either not part of the ipcache or info->tunnel_endpoint
   is not set, then we have no connectivity to such address. The v6 behavior
   is slightly different compared to v4 here in the sense that we only do
   a tunnel map lookup in case /96 prefix matches for the address. Looking
   at the history, the cluster prefix match had been removed for v4 back
   in 661062fec8a1 ("bpf: Remove cluster identity") but w/o corresponding
   v6 removal. Meaning, we never attempt to push the packet through the
   tunnel, and therefore we have no connectivity.

In order to fix both issues, we 1) remove the ipv6_match_prefix_96() in
order to get the same tunnel lookup behavior as with v4 and therefore
allow reachablility for such addresses, and 2) we NAT cilium_host addresses
to the ROUTER_IP when pushed through the tunnel. The ROUTER_IP already comes
from a Cilium node's allocation pool, but adding it to the cilium_host
interface is not an option since it's going to be scoped globally and is
therefore public.

Given the node's ROUTER_IP is already a reserved address, we simply NAT
the cilium_host source IP whenever we redirect to the tunnel /and/ whenever
we receive traffic via tunnel to our own node's ROUTER_IP. For the latter
we then reverse NAT back into HOST_IP. This mechanism ensures that traffic
will stay within the tunnel. Any requests to a remote Cilium node's public
address won't be NATed. The NodePort overlay SNAT needed to be moved into
its own tail call as otherwise we blow up the 4k insns limit.

Fixes: #8654
Fixes: 661062fec8a1 ("bpf: Remove cluster identity")
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -597,6 +600,8 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
 		ret = encap_and_redirect_lxc(skb, tunnel_endpoint, encrypt_key, &key, SECLABEL, monitor);
+		if (!revalidate_data(skb, &data, &data_end, &ip4))
+			return DROP_INVALID;
```
The IPv4 path here differs from the IPv6 path in the ordering of the new `revalidate_data()` call vs. encap return value checks. Is this significant?

joestringer reviewed (dismissed):
Looks good, one minor nit below which I think we can fix to make the ipv4/ipv6 cases more consistent.

Might need some special logic for encryption mode cases, given the testsuite failure. /cc @jrfastab

borkmann committed:
k8s: replace NodePort frontend cilium_host IP with router addr

In case of tunneling, add the internal v6 address which is part of the
cluster cidr, not the public address which is already covered through
the GetNodePortIPv6().

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

jrfastab commented on bpf/lib/encap.h:
```diff
@@ -64,6 +64,63 @@ encap_and_redirect_ipsec(struct __sk_buff *skb, __u32 tunnel_endpoint, __u8 key,
+	}
+	sum = csum_diff(which, 16, &host_ip, 16, 0);
+	csum_l4_offset_and_flags(nexthdr, &csum);
+	if (skb_store_bytes(skb, noff, &host_ip, 16, 0) < 0)
```
would be nice to have some define for '16' number here but not critical could be done later.

borkmann committed:
bpf: fix routing of cilium_host router ip and health in v6 tunnel mode

Even though the v4 cilium_host as well as health IPs are properly set up
with 192.168.1.120 as tunnel endpoint as well as remote v4+v6 pod addresses,
the cilium_host router IP is not and therefore attempted to be routed
publicly which is incorrect. Same for health by the way. Noticed as BPF
based nodeport replies in v6 via tunnel to cilium_host were correctly
NATed but routed via public interface instead of cilium_vxlan:

Tunnel map:

  # ./cilium/cilium bpf tunnel list
  TUNNEL           VALUE
  10.21.0.0:0      192.168.1.120:0
  2001:db1:3:::0   192.168.1.120:0

Before:

  # ./cilium/cilium bpf ipcache list
  IP PREFIX/ADDRESS                            IDENTITY
  2001:db1:3::c8a4/128                         4 0 0.0.0.0                <-- Remote node's health IPv6 (BUG: no tunnel entry)
  fe80::3090:86ff:fe32:2e11/128                1 0 0.0.0.0
  10.11.39.69/32                               1 0 0.0.0.0
  10.11.118.73/32                              4 0 0.0.0.0
  10.21.251.113/32                             1 0 192.168.1.120          <-- Remote node's cilium_host IPv4
  192.168.1.125/32                             1 0 0.0.0.0
  2001:db1:3::f9a/128                          16777217 0 192.168.1.120   <-- Remote pod IPv6 address
  2001:db1:3::6e04/128                         1 0 0.0.0.0                <-- Remote node's cilium_host [router] IPv6 (BUG: no tunnel entry)
  ::/0                                         2 0 0.0.0.0
  2001:db8:5::9c23/128                         4 0 0.0.0.0
  2a02:120b:2c12:c120:ec4:7aff:fe03:f925/128   1 0 0.0.0.0
  10.21.157.26/32                              16777217 0 192.168.1.120   <-- Remote pod IPv4 address
  10.21.184.75/32                              4 0 192.168.1.120          <-- Remote node's health IPv4
  0.0.0.0/0                                    2 0 0.0.0.0
  2001:db8:5::a36a/128                         1 0 0.0.0.0

After:

  # ./cilium/cilium bpf ipcache list
  IP PREFIX/ADDRESS                            IDENTITY
  10.11.39.69/32                               1 0 0.0.0.0
  10.21.184.75/32                              4 0 192.168.1.120
  192.168.1.125/32                             1 0 0.0.0.0
  2001:db1:3::6e04/128                         1 0 192.168.1.120          <-- Remote node's cilium_host [router] IPv6 via tunnel now
  2001:db1:3::c8a4/128                         4 0 192.168.1.120          <-- Remote node's health IPv6 via tunnel now
  2001:db8:5::9240/128                         4 0 0.0.0.0
  2a02:120b:2c12:c120:ec4:7aff:fe03:f925/128   1 0 0.0.0.0
  fe80::3090:86ff:fe32:2e11/128                1 0 0.0.0.0
  10.11.172.94/32                              4 0 0.0.0.0
  0.0.0.0/0                                    2 0 0.0.0.0
  2001:db1:3::f9a/128                          16777217 0 192.168.1.120
  2001:db8:5::a377/128                         4 0 0.0.0.0
  10.21.251.113/32                             1 0 192.168.1.120
  10.21.157.26/32                              16777217 0 192.168.1.120
  2001:db8:5::a36a/128                         1 0 0.0.0.0
  ::/0                                         2 0 0.0.0.0
  10.11.115.69/32                              4 0 0.0.0.0

BPF NodePort request via cilium_vxlan where backend is local working now:

  23:27:40.224225 IP6 2001:db1:3::6e04.46593 > 2001:db8:5::a36a.30000: Flags [S], seq 1928315317, win 27800, options [mss 1390,sackOK,TS val 1737072991 ecr 0,nop,wscale 7], length 0
  23:27:40.224263 IP6 2001:db8:5::a36a.30000 > 2001:db1:3::6e04.46593: Flags [S.], seq 1813566207, ack 1928315318, win 64260, options [mss 1440,sackOK,TS val 1976342860 ecr 1737072991,nop,wscale 7], length 0
  23:27:40.224501 IP6 2001:db1:3::6e04.46593 > 2001:db8:5::a36a.30000: Flags [.], ack 1, win 218, options [nop,nop,TS val 1737072992 ecr 1976342860], length 0
  [...]

Fixes: f6300d90d09a ("node: Update ipcache entries independent of node update source")
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb reviewed (approved):
LGTM for Go parts.
